### ISSUE TITLE
XEP-0384: OMEMO Encryption 0.4.0

### ISSUE THREAD

**Issue by Neustradamus** [gkdr/libomemo#25]
Can you update the XEP-0384: OMEMO Encryption support to 0.4.0?

- http://www.aptest.com/standards/htmldiff/htmldiff.pl?oldfile=https://xmpp.org/extensions/attic/xep-0384-0.3.0.html&newfile=https://xmpp.org/extensions/attic/xep-0384-0.4.0.html

**Comment by gkdr** [comment]
as far as i know there is no client implementing this. am i wrong?

**Comment by selurvedu** [comment]
Yes

**Comment by Neustradamus** [comment]
Here: https://github.com/dino/dino/tree/feature/omemo1

**Comment by gkdr** [comment]
nice, that means the `libsignal-protocol-c` fork is functional as well?

**Comment by Neustradamus** [comment]
@gkdr: First, I can not confirm that current libomemo supports a perfect OMEMO 0.3.0, only you can confirm it.

About OMEMO 0.4.0+, I am not sure that the omemo1 branch of Dino is up-to-date.
It is important to create a new libomemo branch for OMEMO 0.4.0+ or better, I think, a new repo, for example "libomemo-v2" or other...

@mar-v-in: Please talk with @gkdr.

Thanks in advance.

**Comment by mar-v-in** [comment]
[libomemo-c](https://github.com/dino/libomemo-c) should work, but afaik it was not thoroughly tested with for compatibility any other implementation yet. When operating based on libomemo-c, you can implement omemo 0.3.0 and 0.4.0+ at the same time so that your client remains fully backwards-compatible. This is what is implemented in [omemo1 branch](https://github.com/dino/dino/tree/feature/omemo1), but again, nothing of this was thoroughly tested and you'd be the first client to implement this in production.

**Comment by gkdr** [comment]
oh, nice, thanks for that! the changes look like i can probably plug it into my signal-protocol-client-code with little adaptations. i can also start implementing the pure xml stuff here, but will think about how to include the forked signal protocol since it's now part of the omemo spec.

**Comment by hades** [comment]
Is this fixed now (after 6f153c086b1b5618fca347f420fe50a8f73e14d5 and further changes last year)?

**Comment by Neustradamus** [comment]
@hades: Sorry for the delay, it does not support OMEMO 0.8.3, it is only 0.3.0.

And "XEP-0454: OMEMO Media sharing" is missing too:
- https://xmpp.org/extensions/xep-0454.html

Maybe some guys can look it?

All links in description.

cc: @root-hardenedvault, @hartwork, @fortysixandtwo.

**Comment by Neustradamus** [comment]
I think it is time to update the lib with latest XEP-0384 version.

People would like the support of the latest version.

**Comment by gkdr** [comment]
@Neustradamus I agree the newest version would be nice. As mentioned above, this would require using the omemo-specific crypto in a completely different place than here (as this module only dealt with the XML protocol and left the crypto to a libsignal-protocol-c wrapper). Not sure what the newest versions add beyond that, is there a good implementation guide somewhere?

**Comment by Neustradamus** [comment]
At this time, it is: 0.9.0 (2025-04-07), previous was 0.8.3 (2022-01-18):
- https://xmpp.org/extensions/xep-0384.html

**Comment by Neustradamus** [comment]
For informations, there are:
- https://github.com/JustNevi/cmemo
- https://github.com/mierenhoop/picomemo
- https://github.com/dino/libomemo-c

**Comment by Neustradamus** [comment]
At this time, it is: 0.9.1 (2026-04-06), previous was 0.9.0 (2025-04-07):
- https://xmpp.org/extensions/xep-0384.html
